Resolve fully qualified names

hubertp opened the pull request:
### Pull Request Description

Added a separate pass, `FullyQualifiedNames`, that partially resolves fully qualified names. The pass only resolves the library part of the name and replaces it with a reference to the `Main` module.

There are 2 scenarios that could be potentially:
1) the code uses a fully qualified name to a component that has been
   parsed/compiled
2) the code uses a fully qualified name to a component that has **not** be
   imported

For the former case, it is sufficient to just check `PackageRepository` for the presence of the library name.
In the latter we have to ensure that the library has been already parsed and all its imports are resolved. That would require the reference to `Compiler` in the `FullyQualifiedNames` pass, which could then trigger a full compilation for missing library. Since it has some undesired consequences (tracking of dependencies becomes rather complex) we decided to exclude that scenario until it is really needed.

### Important Notes

With this change, one can use a fully qualified name directly.
e.g. 
```
import Standard.Base
main =
  Standard.Base.IO.println "Hello world!"
```

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Resolve fully qualified names

Added a separate pass, `FullyQualifiedNames`, that partially resolves
fully qualified names. The pass only resolves the library part of the
name and replaces it with a reference to the `Main` module.

There are 2 scenarios covered in this change:
1) the code uses a fully qualified name to a component that has been
   parsed/compiled
2) the code uses a fully qualified name to a component that has not be
   imported

For the former case, it is sufficient to just check `PackageRepository`
for the presence of the library name. In the latter case we have to
ensure that the library has been already parsed and all its imports are
resolved. Hence the reference to `Compiler` in the `FullyQualifiedNames`
pass.

hubertp commented on engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/FullyQualifiedNames.scala:
```diff
@@ -0,0 +1,336 @@
+            s"${libName.toString}.${Imports.mainModuleName.name}"
+          )
+          .map { m =>
+            if (m.getIr == null) {
```
This change is probably the most controversial as it means triggering a full compilation of the library.

We could limit the scope of this change and issue an error here instead. In that case, we wouldn't be able to use a fully qualified names unless someone imported (parts of) it before. So far, that would be extremely rare since regular programs are rather limited without stdlib.

radeusgd replied:
I like limiting FQNs to just imported libraries. I'm not sure if that's what we want as at this point they are less useful (wasn't their whole point the fact that you can use them without imports?).

But I think, especially as in Enso the imports actually _define_ which packages a project relies on, it's good to have one place where this is define - i.e. the imports which are relatively easy to glance. If we add FQNs within the code, it will be much harder to find out which libraries are _actually_ imported.

So I would vote for indeed only allowing to use an FQN on an imported library, but I understand that we may need to do that differently due to other requirements, so I'm also not opposed to allowing FQNs to load libraries if that is needed.

hubertp replied:
That can be reduced fairly easy by reporting the error in the condition rather than triggering `compiler.run`.
That would mean however that
```
main =
  Standard.Base.IO.println "Hello world!"
```
needs to be
```
import Standard.Base
main =
  Standard.Base.IO.println "Hello world!"
```
at least.

radeusgd replied:
I like this, we can make the error quite readable ("FQN used, but its library is not imported, please add the missing `import Foo.Bar` import" - we can even include the code in the error message).
The IDE should also be able to handle adding the proper import. It will also be easier for the IDE too to track stuff if all library dependencies are only in the imports and not scattered around the whole codebase.

radeusgd replied:
But I'm not the one who was requesting making FQN's work, so probably needs their opinion too to make sure such behaviour would be ok. I'm mostly looking from the perspective of making it easy to track the dependencies.

4e6 replied:
```py
main =
  Standard.Base.IO.println "Hello world!"
```
I think we can schedule compilation of `Stdandard.Base.IO`, and the imports resolver algorithm will figure out all the required dependencies. Although, it can be implemented as an optimization in the future. I feel it's ok to trigger the compilation of `Standard.Base` in this case.

JaroslavTulach replied:
The IDE currently `from Standard.Base import all`. Then the whole _standard base library_ should be compiled and one could use FQN to refer to any of its types, right? If so, then we may not need to cyclic dependency from `ModuleContext` to `Compiler` probably. But I am not a language designer.

hubertp replied:
So I feel like it would be safer to limit this PR on purpose as there are a lot of unknowns here. We could come back to this if there is really a demand for import-less FQNs.

radeusgd replied:
After all, I think the IDE is supposed to have a whole import-management capability, so this seems reasonable, especially if we report a friendly error

hubertp replied:
done

radeusgd reviewed (commented):
We have a [`CompilerBasedDependencyExtractor`](https://github.com/enso-org/enso/blob/46c31bb9a52008a741f8ff9b71d8395374882e67/engine/language-server/src/main/scala/org/enso/languageserver/libraries/CompilerBasedDependencyExtractor.scala) that is used to extract dependencies between libraries based on imports.

It is meant to be used to track dependencies between our libs, it is currently not used too heavily, although it is ran as part of generating a manifest for a package.

Since this PR introduces a change which makes libraries loaded not only by imports but also by qualified names, I'd like to ask:

1. If we can make sure that this dependency extractor will also be able to detect which libraries are brought in as part of this pass? If so, it would be good to add a test extracting dependencies of `Test_Fully_Qualified_Name` verifying this works.
  1a. Alternatively, maybe we could actually make a pass which desugars the FQN into a regular reference + adding a fake import to the module? What do you think? Maybe then we could avoid modifications in the dependency extractor. It seems quite valuable to have a single place to find dependencies (the list of imports).

2. Since this is not necessarily top priority, we may mark this for later, but I think we should create a Pivotal issue to tackle this at some point, to avoid figuring this out in the future by a bug where library dependencies are not tracked correctly.

hubertp commented:
> We have a [`CompilerBasedDependencyExtractor`](https://github.com/enso-org/enso/blob/46c31bb9a52008a741f8ff9b71d8395374882e67/engine/language-server/src/main/scala/org/enso/languageserver/libraries/CompilerBasedDependencyExtractor.scala) that is used to extract dependencies between libraries based on imports.

Wasn't aware of that. Thanks @radeusgd

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -322,7 +322,8 @@ class Compiler(
           freshNameSupply = Some(freshNameSupply),
-          compilerConfig  = config
+          compilerConfig  = config,
+          compiler        = Some(this)
```
Now `ModuleContext` needs a compiler. I see it as a cyclic dependency and I am afraid of it.

hubertp replied:
Yes, changed that to `PackageRepository` since we no longer trigger compilation. This is not ideal but due to the way passes and metadata are modeled, passes **have to** be defined as `case object` rather than `case class` and additional info has to be passed via context.

Ideally we would rewrite that since in situations like this it feels awkward but it's a much bigger task.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/context/InlineContext.scala:
```diff
@@ -63,7 +65,8 @@ object InlineContext {
       passConfiguration = moduleContext.passConfiguration,
-      compilerConfig    = moduleContext.compilerConfig
+      compilerConfig    = moduleContext.compilerConfig,
+      compiler          = moduleContext.compiler
```
Why one needs `compilerConfig` when one has a pointer to `compiler`?

JaroslavTulach reviewed (approved):
I have slight preference to smaller change - e.g. without invoking the compiler - if that works when the IDE imports `from Standard.Base import all`.

hubertp committed:
Limit Fully Qualified Names

Don't trigger compilation of library late in the pipeline as this may
have undesired consequences for dependency resolutions.
This will be a known limitation of FQN.

hubertp committed:
Replace compiler ref with package repo one

We are no longer triggering compilation so that is not necessary.
Unfortunately, due to the way passes and metadata are currently modeled
we cannot pass `PackageRepository` directly to the pass and we have to
do it via context.

hubertp committed:
changelog

radeusgd commented on engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/FullyQualifiedNames.scala:
```diff
@@ -0,0 +1,358 @@
+    optPkgRepo: Option[PackageRepository]
+  ): Either[IR.Expression, Option[FQNResolution]] = {
+    optPkgRepo
+      .flatMap { pkgRepo =>
+        val libName = LibraryName(thisResolution.namespace, consName.name)
+        pkgRepo.ensurePackageIsLoaded(libName).toOption.flatMap { _ =>
+          pkgRepo
+            .getLoadedModule(
+              s"${libName.toString}.${Imports.mainModuleName.name}"
+            )
+            .map { m =>
+              if (m.getIr == null) {
+                // Limitation of Fully Qualified Names:
+                // If the library has not been imported explicitly, then we won't have
+                // IR for it. Triggering a full compilation at this stage may have
+                // undesired consequences and is therefore prohibited on purpose.
+                Left(
+                  IR.Error.Resolution(
+                    consName,
+                    MissingLibraryImportInFQNError(thisResolution.namespace)
+                  )
+                )
+              } else {
```
If we are not allowing FQNs of unimported modules, I don't think this is the right approach to implement it.

Please keep in mind that `ensurePackageIsLoaded` is potentially very costly operation - it may even _download_ a new library if the requested library is not yet cached.

radeusgd replied:
I appreciate that we want to distinguish 'module was not imported' vs 'module with such path does not exist in that library'.

But without importing (downloading) the library, we cannot really tell the second one reliably.

I would suggest then to add `isPackageLoaded` to the `PackageRepository` which just checks if it's loaded, and then do the following (in pseudocode):

```
if (pkgRepo.isPackageLoaded(libName)) {
   m = pkgRepo.getLoadedModule(s"${libName.toString}.${Imports.mainModuleName.name}")
   // do we still need to check if m.getIR == null? If it was imported it should already be compiled. I don't think at this point this scenario is viable but if it is, we can check
   Right(...)
} else {
   Left(MissingLibraryImport(libName))
}
```

That way we will avoid the costly package load/download.

hubertp replied:
Fair comment. I will make the change

hubertp committed:
Address PR review

radeusgd commented on engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala:
```diff
@@ -7395,6 +7395,13 @@ object IR {
+      case class MissingLibraryImportInFQNError(namespace: String)
+          extends Reason {
+        override def explain(originalName: IR.Name): String =
+          s"Fully qualified name includes a library $namespace.${originalName.name} but import statement is missing."
```
Just a very minor suggestion, feel free to ignore.

```suggestion
          s"Fully qualified name references a library $namespace.${originalName.name} but an import statement for it is missing."
```

IMHO `references` reads better, as `includes` suggests that it _does do the import_ (at least to someone biased with C/C++ terminology). Reference seems a bit more neutral and specific about what is happening.

hubertp committed:
Update engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

hubertp committed:
Adapt test output to the suggestion

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/fqn-resolution-184213601
